In [56]:
from collections import *
class caract:
    def __init__(self,data):
        self.level = 0
        self.cost = int(data[1])
        self.skills = (set(data[3:]))
        
    def __repr__(self):
        return '''LEVEL {} COST {} SKILL {}'''.format(self.level,self.cost,self.skills)
    def bound_skills(self,s):
        data = [(self.level),self.cost,0]
        data.extend(self.skills.intersection(s ))
        return caract(data)
    def __add__(self,o):
        data = [max(self.level,o.level),self.cost+o.cost,0]
        data.extend(self.skills|o.skills)
        a=caract(data)
        return a
    def get_base(self):
        #copy free cost=0
        S= caract([self.level,0,0,]+list(self.skills))
        return S
    def __eq__(self,other):
        return self.skills==other.skills and self.cost==other.cost
    def __hash__(self):
        return hash(tuple(sorted(self.skills)+[self.cost]))
def load_almanaq():
    N=open('almanac.data','r')
    C,F =next(N).split()
    character={}
    for c in range(int(C)):
        line=next(N).strip().split()
        character[line[0]]=caract(line[1:])
                 
    Fus=defaultdict(list)
    for f in range(int(F)):
        line=next(N).strip().split()
        Fus[line[0]].append(line[1:])
    return character,Fus
C,F=load_almanaq()

In [57]:
import operator
from functools import reduce
operator.add

<function _operator.add>

In [58]:
def clean(list_caract):
    SKILLS_COSTS={}
    for i in list_caract:
        SKILLS_COSTS[tuple(sorted(i.skills))]=min(i.cost,SKILLS_COSTS.get(tuple(sorted(i.skills)),999999999))
    ans=[]
    for u,v in SKILLS_COSTS.items():
        ans.append(caract((0,v,0,)+u))
      
    return ans

In [59]:
BASE={c:C[c].get_base() for c in C}
def encarntions_ch(character,bounded_skills,presup):
    if character in encarntions_ch.D:
        return encarntions_ch.D[character]
    ans=set()
    ans.add(C[character])
    for fusion in F[character]:
        for l in list(encarntions_ch(fusion[0],bounded_skills,presup)):
            
            for r in list(encarntions_ch(fusion[1],bounded_skills,presup)):
                fus = (r+l+BASE[character]).bound_skills(bounded_skills)
                if fus.cost<=presup:
                    ans.add( fus )
    
    encarntions_ch.D[character]=clean(ans)
    return ans
encarntions_ch.D ={}


In [60]:
set(['character18', 'pixie']+
 ['character80', 'character87']+
 ['character26', 'character5']+
 ['character38', 'character4']+
 ['character101', 'thor']+
 ['character76', 'character83'])-set( encarntions_ch.D)

{'character101',
 'character18',
 'character26',
 'character38',
 'character4',
 'character5',
 'character76',
 'character80',
 'character83',
 'character87',
 'pixie',
 'thor'}

In [77]:
'character149' in ( in (encarntions_ch.D))

True

In [82]:
for i in F['character176']:
    for k in i:
        if k not in encarntions_ch.D:
            print(k)

character143
character159


In [66]:
for i in (encarntions_ch.D):
    print(i,encarntions_ch.D[i].__len__())

character109 122
character130 192
character78 5
angel 2
character19 1
character76 10
character88 3
character16 1
character31 1
character45 2
character30 1
character129 80
character37 1
character126 110
character46 3
character100 9
character123 105
character63 3
character4 1
character24 1
character52 1
character11 1
character147 166
character1 1
character142 257
character65 12
character112 31
character86 19
character116 64
character140 254
character115 106
character25 1
character61 1
character110 100
character38 2
character40 1
character13 1
character97 5
character79 6
character54 3
character2 1
character89 3
character29 1
character26 2
character59 5
character72 5
character49 2
character5 1
character119 167
character22 1
character64 6
character87 5
character47 1
character80 13
pixie 1
character133 42
character34 2
character8 1
character121 4
character43 1
character114 127
character57 3
character81 3
character84 3
character20 2
character32 1
character58 2
mithra 2
character96 29
characte

In [85]:
def main():
    N=open('submitInput.txt','r')
    with open('submitOutput.txt','w') as f:
        
        for i in range(int(next(N))):
            line=next(N).strip().split()
            
            #f=None
            gold,name,_=line[:3]
            gold=int(gold)
            skill = set(line[3:])
            get=False
            #print(gold,skill,name)
            encarntions_ch.D ={}
            for t in sorted(encarntions_ch(name,skill,gold),key=lambda x:x.cost):
                #print(t)
                if t.cost<=gold and t.skills.issuperset(skill):
                    get=True
                    cost=t.cost
                    main.t=t
                    break
                
            if get:
                print('Case #{}: {}'.format(i+1,cost),)
                print('Case #{}: {}'.format(i+1,cost),file=f)
            else:
                print('Case #{}: {}'.format(i+1,'IMPOSSIBLE'),file=f)
                print('Case #{}: {}'.format(i+1,'IMPOSSIBLE'),)
                
main()                

Case #47: 4220
Case #48: 6200
Case #49: 6540
Case #50: 9160


In [13]:
!head submitInput.txt

50
87000 character97 1 skill13
23500 character125 1 skill68
86140 character118 2 skill78 skill4
97860 character127 3 skill67 ice skill34
11440 character102 3 skill21 wind skill58
39700 character103 3 skill34 skill19 skill8
42000 character118 3 skill13 skill34 skill42
37580 character120 3 skill21 resist_thunder skill61
23200 character138 3 cure wind skill69


encarntions_ch.D

Case #1: IMPOSSIBLE
Case #2: 4110
Case #3: 3350
Case #4: 3220
Case #5: 3820
Case #6: IMPOSSIBLE
Case #7: 4470
Case #8: IMPOSSIBLE
Case #9: 2410
Case #10: 4960
Case #11: 3870
Case #12: 7610
Case #13: 4660
Case #14: 6390
Case #15: 3230
Case #16: 3910
Case #17: IMPOSSIBLE
Case #18: 4870
Case #19: 7000
Case #20: 3580
Case #21: 4360
Case #22: 4660
Case #23: 3850
Case #24: 4510
Case #25: 3520
Case #26: IMPOSSIBLE
Case #27: 4430
Case #28: 6350
Case #29: 4030
Case #30: IMPOSSIBLE
Case #31: 3710
Case #32: 5270
Case #33: 5420
Case #34: IMPOSSIBLE
Case #35: 8360
Case #36: 6240
Case #37: 9980
Case #38: 4600
Case #39: 4030
Case #40: 5450
Case #41: 5620
Case #42: 9840
Case #43: 4880
Case #44: 8790
Case #45: 6530
Case #46: 6310
Case #47: 4220
Case #48: 6200
Case #49: 6540
Case #50: 9160